src/java/org/apache/cassandra/locator/Ec2MetadataServiceConnector.java

```diff
@@ -0,0 +1,234 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.locator;
+
+import java.io.IOException;
+import java.time.Duration;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableMap;
+
+import org.apache.cassandra.exceptions.ConfigurationException;
+import org.apache.cassandra.utils.Clock;
+import org.apache.cassandra.utils.Pair;
+
+import static java.lang.String.format;
+
+public interface Ec2MetadataServiceConnector extends CloudApiCallable
+{
+    String EC2_METADATA_TYPE_PROPERTY = "ec2_metadata_type";
+    String DEFAULT_METADATA_SERVICE_URL = "http://169.254.169.254";
+
+    enum EC2MetadataType
+    {
+        V1("v1"),
+        V2("v2");
+
+        private final String name;
+
+        EC2MetadataType(String name)
+        {
+            this.name = name;
+        }
+
+        @Override
+        public String toString()
+        {
+            return name;
+        }
+    }
+
+    static Ec2MetadataServiceConnector create(SnitchProperties props)
+    {
+        try
+        {
+            EC2MetadataType type = EC2MetadataType.valueOf(props.get(EC2_METADATA_TYPE_PROPERTY, EC2MetadataType.V2.toString()).toUpperCase());
+
+            switch (type)
+            {
+                case V1:
+                    return V1Connector.create(props);
```

Comment:
well, provider could be included in enum, then we would just call `type.create(props)`

Reply:
great idea.